File: dailyTemperatures.java
import java.util.*;
// Approach 1- Brute Force 
class Solution{ // TC O(n^2) SC O(1)
	public int[] dailyTemperatures(int[] temperatures){
		int[] ans= new int[temperatures.length];
		for(int i=0; i<temperatures.length; i++){
			for(int j=i+1; j<temperatures.length; j++){
				if(temperatures[j]>temperatures[i]){
					ans[i]=j-i;
					break;
				}
			}
		}
		return ans;
	}
}

/* Approach 2- Decreasing Monotonic Stack

Monotonic stacks are a good option when a problem involves 
comparing the size of numeric elements, with their order being relevant.
*/

class Solution1{ // TC O(n) SC O(n)
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] answer = new int[n];
        Deque<Integer> stack = new ArrayDeque<>();
        
        for (int currDay = 0; currDay < n; currDay++) {
            int currentTemp = temperatures[currDay];
            // Pop until the current day's temperature is not
            // warmer than the temperature at the top of the stack
            while (!stack.isEmpty() && temperatures[stack.peek()] < currentTemp) {
                int prevDay = stack.pop();
                answer[prevDay] = currDay - prevDay;
            }
            stack.push(currDay);
        }
        
        return answer;
    }
}

// Approach 3- Array, Optimizied space
class Solution2{ // TC O(n) SC O(1)
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int hottest = 0;
        int answer[] = new int[n];
        
        for (int currDay = n - 1; currDay >= 0; currDay--) {
            int currentTemp = temperatures[currDay];
            if (currentTemp >= hottest) {
                hottest = currentTemp;
                continue;
            }
            
            int days = 1;
            while (temperatures[currDay + days] <= currentTemp) {
                // Use information from answer to search for the next warmer day
                days += answer[currDay + days];
            }
            answer[currDay] = days;
        }
        
        return answer;
    }
}

class dailyTemperatures{
	public static void main(String[] args) {
		int[] temperatures={73,74,75,71,69,72,76,73};
		Solution s= new Solution();
		System.out.println(Arrays.toString(s.dailyTemperatures(temperatures)));
		Solution1 s1= new Solution1();
		System.out.println(Arrays.toString(s1.dailyTemperatures(temperatures)));
	}
}

File: dailyTemperatures.py
from collections import deque
# Approach 1- Brute Force 
def dailyTemperatures(temperatures): # TC O(n^2) // SC O(1)
	ans=[0]*len(temperatures)
	for i in range(len(temperatures)):
		for j in range(i+1,len(temperatures)):
			if temperatures[j]>temperatures[i]:
				ans[i]=j-i 
				break 

	return ans 

# Approach 2- Decreasing Monotonic Stack
def dailyTemperatures1(temperatures): # TC O(n) // SC O(n)
	deq=deque()
	ans=[0]*len(temperatures)
	for i in range(len(temperatures)):
		currTemp=temperatures[i]
		while deq and temperatures[deq[-1]]<currTemp:
			prevDay=deq.pop()
			ans[prevDay]=i-prevDay 
		deq.append(i)

	return ans

# Approach 3- Array, Optimizied space
def dailyTemperatures2(temperatures): # TC O(n) // SC O(1)
	n = len(temperatures)
	hottest = 0
	answer = [0] * n

	for curr_day in range(n - 1, -1, -1):
	    current_temp = temperatures[curr_day]
	    if current_temp >= hottest:
	        hottest = current_temp
	        continue
	    
	    days = 1
	    while temperatures[curr_day + days] <= current_temp:
	        # Use information from answer to search for the next warmer day
	        days += answer[curr_day + days]
	    answer[curr_day] = days

	return answer 

print(dailyTemperatures([73,74,75,71,69,72,76,73]))
print(dailyTemperatures1([73,74,75,71,69,72,76,73]))
print(dailyTemperatures2([73,74,75,71,69,72,76,73]))